guides/inventing-service.md

```diff
@@ -0,0 +1,856 @@
+someone on the internet. Our `Timeout` wrapper is actually useful here as well.
+`JsonContentType` probably isn't since its not the clients job to set response
+headers.
+
+Since our `Handler` trait is useful for defining both servers and clients,
+`Handler` probably isn't an appropriate name anymore. A client doesn't handle an
+request, it sends it to a server and the server handles it. Lets instead call
+our trait `Service`:
+
+```rust
+trait Service<Request> {
+    type Response;
+    type Error;
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    fn call(&mut self, request: Request) -> Self::Future;
+}
+```
+
+This is actually _almost_ the `Service` trait defined in Tower. If you've been
+able to follow along until this point you now understand most of Tower. Besides
+the `Service` trait Tower also provides several utilities that implement
+`Service` by wrapping some other type that also implements `Service`. Exactly
+like we did with `Timeout` and `JsonContentType`. Things can be composed in ways
+similar to what we've just did.
+
+Some example services provided by Tower:
+
+- `Timeout` - This is pretty much identical to the timeout we have built.
+- `Retry` - To automatically retry failed requests.
+- `RateLimit` - To limit the number of requests a service can handle over a
+  period of time.
+
+Types like `Timeout` and `JsonContentType` are typically called "middlewares"
+since they wrap another `Service` and transform the request or response in some
+way. Whereas types like `RequestHandler` is typically called "leaf services"
+since it sits at the leaves of a tree of nested services. The actual responses
+are normally produced in leaf services and modified by middlewares.
+
+The only thing left to talk about is "backpressure" and `poll_ready`.
+
+## Backpressure
+
+Imagine you wanted to write a rate limit middleware that wraps a `Service` and
+puts a limit on the maximum number of concurrent requests the underlying service
+can will receive. This would be useful if you had some service that had a hard
+upper limit on the amount of load it could handle.
+
+With our current `Service` trait we don't really have a good way to implement
+something like that. We could try:
+
+```rust
+impl<R, T> Service<R> for ConcurrencyLimit<T> {
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of running requests
+        //    currently.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait a bit and check the capacity
+        //    again.
+    }
+}
+```
+
+If there is no capacity left we just have to wait a bit and then check again. So
+if the service is receiving a lot of requests you can imagine waiting for a long
+time.  It would also be nice if we had a way to tell the caller that there was
+no more capacity left. Maybe they wanted to react to that by calling another
+service instead.
```

Comment:
I'm not quite sure this argument is strong enough. Rather than "waiting for some random amount of time" one could await on an async semaphore. However that merges "driving a service readiness" and "calling the service" so `poll_ready` is still better. I'm just not quite sure this section drives that point home.

Reply:
Hopefully 8283ecaec7446 makes things a bit more clear.